packages/babel-helpers/src/helpers/wrapRegExp.js

```diff
@@ -0,0 +1,65 @@
+/* @minVersion 7.2.6 */
+
+import setPrototypeOf from "setPrototypeOf";
+import inherits from "inherits";
+
+export default function _wrapRegExp() {
+  _wrapRegExp = function (re, groups) {
+    return new BabelRegExp(re, undefined, groups);
+  };
+
+  var _super = RegExp.prototype;
+  var _groups = new WeakMap();
+
+  function BabelRegExp(re, flags, groups) {
+    var _this = new RegExp(re, flags);
+    // if the regex is recreated with 'g' flag
+    _groups.set(_this, groups || _groups.get(re));
+    return setPrototypeOf(_this, BabelRegExp.prototype);
+  }
+  inherits(BabelRegExp, RegExp);
+
+  BabelRegExp.prototype.exec = function (str) {
+    var result = _super.exec.call(this, str);
+    if (result) result.groups = buildGroups(result, this);
+    return result;
+  };
+  BabelRegExp.prototype[Symbol.replace] = function (str, substitution) {
+    if (typeof substitution === "string") {
+      var groups = _groups.get(this);
+      return _super[Symbol.replace].call(
+        this,
+        str,
+        substitution.replace(/\$<([^>]+)>/g, function (_, name) {
+          return "$" + groups[name];
+        })
+      );
+    } else if (typeof substitution === "function") {
+      var _this = this;
+      return _super[Symbol.replace].call(this, str, function () {
+        var args = [];
+        args.push.apply(args, arguments);
+        if (typeof args[args.length - 1] !== "object") {
```

Comment:
I think this is incorrect. Because we transform `(?<year>\d{4})` to `([0-9]{4})`, if there is a object here, it won't include the `year` group. I think we ned to pop and push if there's an object, and only push if there's not.

Reply:
The object here is not the one internally built by the engine, but it's the `.groups` property returned by our `exec()` (which defined `.groups.year`.

Reply:
You can verify it with
```js
class MyRegExp extends RegExp {
  exec() {
    return Object.assign(["a", "b"], { groups: { name: "c" } });
  }
  [Symbol.replace](str, subst) {
    super[Symbol.replace](str, (...args) => {
      console.log(args);
      return "_";
    });
  }
}
"abc".replace(new MyRegExp("b"));
```

Reply:
Crazy, I didn't realize it was accessing the `.groups` directly from `result`.

Reply:
One more perf reason that makes implementers hate regexp subclasses :stuck_out_tongue: